tools/lib/dependencies.js

```diff
@@ -0,0 +1,103 @@
+const MAX_REORDER_ATTEMPTS = 10
+
+/**
+ * Reorders languages, moving dependencies above the languages
+ * that depend on them.
+ *
+ * @param {array<Language>} languages list of languages to be reordered
+ * @returns {array<Language>} ordered list of languages
+*/
+
+const reorderDependencies = (languages) => {
```

Comment:
This function is not evidently correct, so I'd like to have a test for it.
(I would trust in building a tree of dependencies and then pushing out languages in breadth-first order.)

Reply:
For building, I think the proof is in the results.  I'm not sure how I feel about writing tests for the build system (we don't have any presently).  Do you think the algorithm isn't self-evident?  It's really just a bubble sort.

If you insist I'd rather spend my time making the load order irrelevant and eliminate this as an issue rather than write tests... This is my least favorite code but I just wanted something simple that I knew worked.

Reply:
The high level (meta) test for the build system should be that our built deliverables actually work, and we already have tests for that (many for node, and a few for browser).  And CDN is really just browser + separate language modules.

Reply:
> Do you think the algorithm isn't self-evident? It's really just a dumb bubble sort.

It doesn't look self-evident to me. It looks like it can fail when depth of dependencies is more then MAX_REORDER_ATTEMPTS. I understand it's unlikely (probably even impossible), but I'd prefer more robust implementation.
E. g. first goes the languages which require nothing, then languages which require already processed languages, and so on. Repeat the process until the list of processed languages stabilizes. All languages that remain unprocessed are known to have recursive dependencies.

Reply:
You want to take a pass at it?  I just wanted something basic and functional. :-)

>  It looks like it can fail when depth of dependencies is more then MAX_REORDER_ATTEMPTS

Maybe, and then it fails with a nice error - that's intentional. :-)  But I think that very, very unlikely to happen.  It really would only happen in case of a circular dependency, which is what the error says.

Reply:
I guess now I'll take a few minutes nd look into removing the whole dependency ordering issue in the first place.

Reply:
npm to the rescue: 

```js
  let resolver = new DependencyResolver();
  for (let lang of languages) {
    resolver.add(lang.name);
    for (let required of lang.requires) {
      resolver.setDependency(lang.name, required);
    }
  }
```

Reply:
I did first try to just remove the whole ordering thing but the tricky problem is aliases... we don't know which aliases map to what until AFTER we've truly registered the languages...

Reply:
I like it better this way.

> but the tricky problem is aliases
Do you mean specifying aliases in a build command (instead of "real" names)? AFAIK it doesn't work now, and I don't think we need to support it.

Reply:
> Do you mean specifying aliases in a build command (instead of "real" names)? AFAIK it doesn't work now, and I don't think we need to support it.

No... I mean that if getLanguage only worked on non-aliases it'd be easy to defer the registration... but since you can pass aliases, now there is no way of knowing which language actually needs to be loaded (it might be an alias).

And you can't just load them all the first time it's called because getLanguage can be called recursively... so then it goes into an endless recursive loop.

Possibly it's fixable, but it wasn't the easy task I thought, so I just went and found a dependency resolver - and it works great. :)